broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/PublishMessageProcessor.java

```diff
@@ -74,35 +75,59 @@ public void processRecord(
       responseWriter.writeRejectionOnCommand(command, RejectionType.BAD_VALUE, rejectionReason);
 
     } else {
-      final TypedBatchWriter batchWriter = streamWriter.newBatch();
-      final long key = batchWriter.addNewEvent(MessageIntent.PUBLISHED, command.getValue());
-      responseWriter.writeEventOnCommand(key, MessageIntent.PUBLISHED, command.getValue(), command);
+      handleNewMessage(command, responseWriter, streamWriter, sideEffect);
+    }
+  }
 
-      matchingSubscriptions =
-          messageStateController.findSubscriptions(
-              messageRecord.getName(), messageRecord.getCorrelationKey());
+  private void handleNewMessage(
+      final TypedRecord<MessageRecord> command,
+      final TypedResponseWriter responseWriter,
+      final TypedStreamWriter streamWriter,
+      final Consumer<SideEffectProducer> sideEffect) {
 
-      for (final MessageSubscription sub : matchingSubscriptions) {
-        sub.setMessagePayload(messageRecord.getPayload());
-      }
+    final TypedBatchWriter batchWriter = streamWriter.newBatch();
+    final long key = batchWriter.addNewEvent(MessageIntent.PUBLISHED, command.getValue());
+    responseWriter.writeEventOnCommand(key, MessageIntent.PUBLISHED, command.getValue(), command);
+
+    // correlate the message only once per workflow instance
+    // - will be improved by #1421
+    matchingSubscriptions =
+        messageStateController
+            .findSubscriptions(messageRecord.getName(), messageRecord.getCorrelationKey())
+            .stream()
+            .collect(Collectors.groupingBy(MessageSubscription::getWorkflowInstanceKey))
+            .values()
+            .stream()
+            .map(sub -> sub.get(0))
+            .collect(Collectors.toList());
+
+    for (final MessageSubscription sub : matchingSubscriptions) {
+      sub.setMessagePayload(messageRecord.getPayload());
 
-      sideEffect.accept(this::correlateMessage);
-
-      if (messageRecord.getTimeToLive() > 0L) {
-        final Message message =
-            new Message(
-                key,
-                messageRecord.getName(),
-                messageRecord.getCorrelationKey(),
-                messageRecord.getPayload(),
-                messageRecord.getMessageId(),
-                messageRecord.getTimeToLive());
-        messageStateController.put(message);
-
-      } else {
-        // don't add the message to the store to avoid that it can be correlated afterwards
-        batchWriter.addFollowUpEvent(key, MessageIntent.DELETED, messageRecord);
+      messageStateController.updateCommandSentTime(sub);
```

Comment:
but the sending is not guaranteed at this point?

Reply:
Correct. It could be that the command is sent later. However, I need to remember that I need to send the command.  I want to introduce a subscription life cycle with #1421. `OPENED -> [ CORRELATING (PENDING) -> CORRELATED ] | -> CLOSED`